src/frame.rs

```diff
@@ -0,0 +1,392 @@
+    }
+
+    /// Decode frame data
+    pub fn decode<T>(&self, offset: usize, length: usize) -> T
+    where
+        T: PrimInt + Unsigned,
+        u64: BitRange<T>,
+    {
+        assert!(
+            (offset + length) <= self.data_length * 8,
+            "Not enough data available"
+        );
+        assert!(length <= size_of::<T>() * 8, "Output type not big enough");
+
+        let num = LittleEndian::read_u64(&self.buffer[0..8]);
+        num.bit_range(offset + length - 1, offset)
+    }
+
+    /// Get the checksum from the frame
+    pub fn get_checksum(&self) -> u8 {
+        self.buffer[self.data_length]
+    }
+
+    /// Get the PID from the frame
+    pub fn get_pid(&self) -> PID {
+        self.pid
+    }
+
+    /// Get the serialized bytes to write to the driver
+    pub fn get_data_with_checksum(&self) -> &[u8] {
+        &self.buffer[0..=self.data_length]
+    }
+}
+
+/// Implements the transport layer of LIN. The units that are transported in a transport layer
+/// frame are called PDU (Packet Data Unit)
+pub mod transport {
+    use super::{Frame, PID};
+
+    /// NAD is the address of the slave node being addressed in a request, i.e. only slave nodes
+    /// have an address. NAD is also used to indicate the source of a response.
+    #[repr(transparent)]
+    pub struct NAD(pub u8);
+
+    /// The PCI (Protocol Control Information) contains the transport layer flow control
+    /// information.
+    #[repr(transparent)]
+    pub struct PCI(u8);
+
+    /// Type of the `PCI` byte
+    pub enum PCIType {
+        /// Single Frame
+        SF = 0,
+        /// First Frame. Start of a multi frame message.
+        FF = 1,
+        /// Consecutive Frame.
+        CF = 2,
+        /// Invalid PCIType
+        Invalid,
+    }
+
+    impl PCI {
+        pub fn new_sf(length: u8) -> PCI {
+            assert!(length <= 5, "Maximum length for single frame is 5");
+            PCI(length + 1)
+        }
+
+        pub fn get_type(self) -> PCIType {
+            match self.0 >> 4 {
+                0 => PCIType::SF,
+                1 => PCIType::FF,
+                2 => PCIType::CF,
+                _ => PCIType::Invalid,
+            }
+        }
+
+        pub const fn get_length(self) -> u8 {
+            self.0
+        }
+    }
+
+    /// The Service Identifier (SID) specifies the request that shall be performed by the slave
+    /// node addressed.
+    #[repr(transparent)]
+    pub struct SID(pub u8);
+
+    #[repr(transparent)]
+    /// The Response Service Identifier (RSID) specifies the contents of the response.
+    pub struct RSID(pub u8);
+
+    /// Create a single frame (CF) PDU
+    pub fn create_single_frame(pid: PID, nad: NAD, sid: SID, data: &[u8]) -> Frame {
+        assert!(
+            !data.is_empty() && data.len() <= 5,
+            "A single frame must contain between 0 and 5 bytes"
+        );
+        // If a PDU is not completely filled the unused bytes shall be filled with ones.
```

Comment:
true if you think binary, but I'd replace `with ones` -> `with 0xFF` since you're talking about a byte, a one would be `0x01`

Reply:
This is actually copied from the specs :wink: But I agree `0xFF` would be clearer.

Reply:
Fixed.